### DISCUSSION TITLE
Need help in Snapshot

### DISCUSSION THREAD

**Discussion by wineternity** [milvus-io/milvus#6884] [Q&A and General discussion]
We still using 0.11.x, and not ready to upgrade to milvus 2.0. So I have a question on 0.11.x's snapshot.

We want to seperate node with read node , write node, and use same mysql database,  after the write node build new collection and index, we copy the collection db file to the read node.   But the read node can only read the new collection after docker container restart.  it seems because milvus 0.11.x use Snapshot to store metadata info.

Could you give some way to implement how to make the read node read the new collection immediately without restart.
It is better could get some analysys or help to introduce how to force update snapshot

**Comment by yhmo** [comment]
Basically, snapshots cache metadata in memory, each time when metadata get changed, we create a snapshot in memory to represent current state of metadata.
When milvus server restart, it will rebuild a snapshot in memory, according to current metadata (read from SQLite/MySQL).

**Reply by wineternity** [reply]
So my problem is the write node change the mysql data, but the read node did not create a new snapshot in memory ?

**Reply by wineternity** [reply]
May I ask how milvus  can get notified the metadata is changed and rebuild the snapshot?

**Reply by wineternity** [reply]
And Is there any interface to force reload the snapshot by hand

**Comment by yhmo** [comment]
You say "copy the collection db file to the read node", could you describe your steps.
Correct me if my understood is incorrect:
1. You start a write node and a read node, share same mysql server.  The two nodes use different db path?
2. Create a collection "A", insert some vectors, build index
3. Copy db path of write node into read node. 
4. Restart read node. Did you restart the write node, too?

**Comment by yhmo** [comment]
At step 4, the read node read metadata from mysql, and build a snapshot in memory. You can check the error.log of the read node, is any error in the log?

**Reply by wineternity** [reply]
You say "copy the collection db file to the read node", could you describe your steps.
Correct me if my understood is incorrect:

1. You start a write node and a read node, share same mysql server. The two nodes use different db path?
= 》 yes , we use same mysql data path, they share data
2. Create a collection "A", insert some vectors, build index
 => yes
3. Copy db path of write node into read node.
 => I copy the collection to the read node, like write_node/db/collections/C_2 to read_node/db_collections, as they share the metadata, the collection id will be unique, I think it is safe
4. Restart read node. Did you restart the write node, too?
=> No, only read node need to be restarted. before restart the read node, the collection on only showed at write node,  in read node, it will report error like this
![image](https://user-images.githubusercontent.com/1088645/127459393-5569af5a-ea29-4787-b805-fc8529ca6f8f.png)
      After read node restart, we can read the the new collection.
      But restart cost a lot and will affect other search req to read node, we want to avoid restart,  even change the source code can accept, But I have no idea about it.

**Reply by yhmo** [reply]
Not sure why it says the collection not found. If you can run into debug mode, set break point in the Snapshots::Init() method, to see why the GetCollectionIDsOperation() didn't find the collection:
```
Status
Snapshots::Init(StorePtr store) {
    auto event = std::make_shared<InActiveResourcesGCEvent>();
    EventExecutor::GetInstance().Submit(event, true);
    STATUS_CHECK(event->WaitToFinish());
    auto op = std::make_shared<GetCollectionIDsOperation>();
    STATUS_CHECK((*op)(store));
    auto& collection_ids = op->GetIDs();
    SnapshotHolderPtr holder;
    for (auto& collection_id : collection_ids) {
        STATUS_CHECK(LoadHolder(store, collection_id, holder));
    }
    return Status::OK();
}
```
